### ISSUE TITLE
Flaky-test: DeferredSyncTest.testForceWillAdvanceLacOnlyUpToLastAcknoledgedWrite

### ISSUE THREAD

**Issue by lhotari** [apache/bookkeeper#4200]
DeferredSyncTest.testForceWillAdvanceLacOnlyUpToLastAcknoledgedWrite is flaky. The build times out since the test never finishes. This happens quite often.

```
"main" #1 prio=5 os_prio=0 cpu=1865.56ms elapsed=1477.95s tid=0x00007f34b4028000 nid=0x149f5 waiting on condition  [0x00007f34bcffc000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.22/Native Method)
	- parking to wait for  <0x000000008ad136e8> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@11.0.22/LockSupport.java:234)
	at java.util.concurrent.CompletableFuture$Signaller.block(java.base@11.0.22/CompletableFuture.java:1798)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@11.0.22/ForkJoinPool.java:3128)
	at java.util.concurrent.CompletableFuture.timedGet(java.base@11.0.22/CompletableFuture.java:1871)
	at java.util.concurrent.CompletableFuture.get(java.base@11.0.22/CompletableFuture.java:2027)
	at org.apache.bookkeeper.common.concurrent.FutureUtils.result(FutureUtils.java:80)
	at org.apache.bookkeeper.common.concurrent.FutureUtils.result(FutureUtils.java:61)
	at org.apache.bookkeeper.client.DeferredSyncTest.testForceWillAdvanceLacOnlyUpToLastAcknoledgedWrite(DeferredSyncTest.java:178)
```

Example failure: https://github.com/apache/bookkeeper/actions/runs/7776777094/job/21204429600?pr=4196#step:12:422

<!-- optionally provide the full stacktrace ->
<details>
<summary>Full exception stacktrace</summary>
<code><pre>
"main" #1 prio=5 os_prio=0 cpu=1865.56ms elapsed=1477.95s tid=0x00007f34b4028000 nid=0x149f5 waiting on condition  [0x00007f34bcffc000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.22/Native Method)
	- parking to wait for  <0x000000008ad136e8> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@11.0.22/LockSupport.java:234)
	at java.util.concurrent.CompletableFuture$Signaller.block(java.base@11.0.22/CompletableFuture.java:1798)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@11.0.22/ForkJoinPool.java:3128)
	at java.util.concurrent.CompletableFuture.timedGet(java.base@11.0.22/CompletableFuture.java:1871)
	at java.util.concurrent.CompletableFuture.get(java.base@11.0.22/CompletableFuture.java:2027)
	at org.apache.bookkeeper.common.concurrent.FutureUtils.result(FutureUtils.java:80)
	at org.apache.bookkeeper.common.concurrent.FutureUtils.result(FutureUtils.java:61)
	at org.apache.bookkeeper.client.DeferredSyncTest.testForceWillAdvanceLacOnlyUpToLastAcknoledgedWrite(DeferredSyncTest.java:178)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@11.0.22/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@11.0.22/NativeMethodAccessorImpl.java:62)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@11.0.22/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@11.0.22/Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator$$Lambda$209/0x0000000100152c40.accept(Unknown Source)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:124)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:[428](https://github.com/apache/bookkeeper/actions/runs/7776777094/job/21204429600?pr=4196#step:12:429))
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)

   Locked ownable synchronizers:
	- None
</pre></code>
</details>

**Comment by lhotari** [comment]
@eolivelli Any ideas why this test DeferredSyncTest.testForceWillAdvanceLacOnlyUpToLastAcknoledgedWrite hangs forever?

**Comment by eolivelli** [comment]
Looking at the code I cannot find any obvious problem.
Maybe in `resumeBookieWriteAcks` things are not working properly ? we are missing some callback

we should reproduce the problem locally

**Comment by tmzk1005** [comment]
Maybe a thread synchronization problem :

![2024-03-20_01-25-DeferredSyncTest](https://github.com/apache/bookkeeper/assets/15995517/f0587fdd-ea11-47b9-8837-d8a5a2aca930)

`code-block-1` and  `code-block-3`  runs in main thread, but `code-block-2` and  `code-block-4` runs in another thread, when  the execution order is `1 -> 2 -> 3 -> 4`. callback has no chance to run, so `result(forceResult);` blocked forever.

**Comment by tmzk1005** [comment]
cc @eolivelli, please review my code to see if I fix #4200 correctly. Thanks!
